patches/server/0164-Fix-MC-117075-TE-Unload-Lag-Spike.patch

```diff
@@ -1,42 +1,158 @@
+From: MrPowerGamerBR <git@mrpowergamerbr.com>
+Date: Sun, 26 Nov 2023 13:02:16 -0300
 Subject: [PATCH] Fix MC-117075: TE Unload Lag Spike
 
+We replaced the `blockEntityTickers` list with a custom list based on fastutil's `ObjectArrayList` with a small yet huge change for us: A method that allows us to remove a list of indexes from the list.
 
+This is WAY FASTER than using `removeAll` with a list of entries to be removed, because we don't need to calculate the identity of each block entity to be removed, and we can jump directly to where the search should begin, giving a performance boost for small removals (because we don't need to loop thru the entire list to find what element should be removed) and a performance boost for big removals (no need to calculate the identity of each block entity).
+
+diff --git a/src/main/java/io/papermc/paper/util/BlockEntityTickersList.java b/src/main/java/io/papermc/paper/util/BlockEntityTickersList.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..c82b0f14257e54848fcbde2c0c6ac746d2b8218c
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/util/BlockEntityTickersList.java
+@@ -0,0 +1,105 @@
++package io.papermc.paper.util;
++
++import it.unimi.dsi.fastutil.ints.IntOpenHashSet;
++import it.unimi.dsi.fastutil.objects.ObjectArrayList;
++import net.minecraft.world.level.block.entity.TickingBlockEntity;
++
++import java.util.Arrays;
++import java.util.Collection;
++
++/**
++ * A list for ServerLevel's blockEntityTickers
++ *
++ * This list is behaves identically to ObjectArrayList, but it has an additional method, `removeAllByIndex`, that allows a list of integers to be passed indicating what
++ * indexes should be deleted from the list
++ *
++ * This is faster than using removeAll, since we don't need to compare the identity of each block entity, and faster than looping thru each index manually and deleting with remove,
++ * since we don't need to resize the array every single remove.
++ */
++public final class BlockEntityTickersList extends ObjectArrayList<TickingBlockEntity> {
++    private final IntOpenHashSet toRemove = new IntOpenHashSet();
++    private int startSearchFromIndex = -1;
++
++    /** Creates a new array list with {@link #DEFAULT_INITIAL_CAPACITY} capacity. */
++    public BlockEntityTickersList() {
++        super();
++    }
++
++    /**
++     * Creates a new array list and fills it with a given collection.
++     *
++     * @param c a collection that will be used to fill the array list.
++     */
++    public BlockEntityTickersList(final Collection<? extends TickingBlockEntity> c) {
++        super(c);
++    }
++
++    /**
++     * Marks an entry as removed
++     *
++     * @param index the index of the item on the list to be marked as removed
++     */
++    public void markAsRemoved(final int index) {
++        // The block entities list always loop starting from 0, so we only need to check if the startSearchFromIndex is -1 and that's it
```

Comment:
While true, I don't like assuming this over code that can change or be extended at any time and possibly go by unnoticed, and adding the extra check isn't expensive, same for a bounds check